### ISSUE TITLE
[INFO] SuSE SLES 11 SP4 kernel memory leak in /proc/schedstat

### ISSUE THREAD

**Issue by h0lzi** [firehol/netdata#3726]
Hi, 
probably not netdata fault. But since I'm checking in every direction and maybe someone has the some Problem. 
I have a few SLES 11 SP4 Server (Kernel 3.0.101-108.7-default)
Monitored  with Netdata version v1.8.0-540-gc1fbd003_rolling

and it looks like Unreclaimable Slab Memory is Leaked somehow.
Can 'tpost screenshots but the proc data:
After 100 Days of uptime
I have
SUnreclaim:     42964948 kB

I startet to notice this since it's Oracle Database Servers and they suddenly started swapping.

My guess it's more a leak to do some call in proc or whatever. Will open a support case with Novel/Suse too. 
But maybe I missed a fix in netdata. Since the version I use  quite "old".

Any tips how to debug this would be nice. But it Probably needs a debug Kernel. Which I can't boot :(

**Comment by h0lzi** [comment]
It's a more generic bucket: size-4096. So that's not much help

**Comment by h0lzi** [comment]
Ok the support and I have drilled it down to the postfix plugin that seems to cause it.
https://github.com/firehol/netdata/blob/master/python.d/postfix.chart.py

If I do a while true; do postqueue -p >>/dev/null ; done
I see SUnreclaim raise also. But once stopped it cleans up. 
While called via netdata it seems not to be cleaned up.
I'm not sure why yet. And why it's only happening on my Servers and not in general.

**Comment by ktsaou** [comment]
netdata runs this command every 10 seconds.
If you do:

```sh
while true; do postqueue -p >>/dev/null; sleep 9; done
```

Does it ever cleanup?

Can you find which entries in `/proc/slabinfo` are increased when `postqueue` increases unreclaimable memory?

**Comment by l2isbad** [comment]
I see no `update_every = 10` in the module source file, so it's every 1 second.

**Comment by ktsaou** [comment]
> I see no update_every = 10 in the module source file, so it's every 1 second.

The default is 10:

https://github.com/firehol/netdata/blob/be414540bc5e3730c0cb3a70476db8587db868a7/conf.d/python.d/postfix.conf#L24

@h0lzi I see the command `slabinfo -S` provides a breakdown of the slab, sorted by memory size.
Can you check it?

**Comment by h0lzi** [comment]
Thx for joining in. SuSe support has figured out the problem.
Was not postfix plugin I first thought. Which would have been strage because more people would have that problem. 
It's the cpuidle plugin: https://github.com/firehol/netdata/blob/master/python.d/cpuidle.chart.py
And from that it's /proc/schedstat that's broken:
For reference:

```
++++++++++++++++++++++++++++++++++
New update:

We checked some of those size-4096 slabs and we saw that the content of those slabs resembles the one from /proc/schedstat.
We then checked the schedstat code, and saw this:

static const struct file_operations proc_schedstat_operations = {
       .open    = schedstat_open,
       .read    = seq_read,
       .llseek  = seq_lseek,
       .release = schedstat_release,
};

static int schedstat_open(struct inode *inode, struct file *file)
{
       return seq_open(file, &schedstat_sops);
}

static int schedstat_release(struct inode *inode, struct file *file)
{
       return 0;
};

This is wrong, as when the file gets released, only schedstat_release() gets called, but this does not free the memory allocated by seq_open/seq_read.

So the above should be:

static const struct file_operations proc_schedstat_operations = {
       .open    = schedstat_open,
       .read    = seq_read,
       .llseek  = seq_lseek,
-       .release = schedstat_release,
+       .release = seq_release,
};
 ++++++++++++++++++++++++++++++++++

```
I'm not sure if the is present in e.g. SLES 12 as I don't have a Test Server.
It's not present in RHEL 7.4+
So I'm not sure if the is an "old" Kernel Bug SuSe did not pull or own code etc.
I have no time to check that.

As this happens only on hardware because in VM's there are no cpu idle stats the plugin is disabled. and not many people are running Netdata on SuSe it was never noticed until now I guess.
So no netdata bug I will close this. But maybe somebody will run into this also. So having this here might still help.

**Comment by l2isbad** [comment]
Nice finding @h0lzi :+1:

**Comment by h0lzi** [comment]
Fixed with kernel-default-devel-3.0.101-108.60.1.x86_64
